File: graze/graze-custom-node-manager/graze_client.py
#!/usr/bin/env python3
import requests
import json
import base64
from typing import Optional, Dict, Any

class GrazeClient:
    def __init__(self, session_cookie: Optional[str] = None):
        self.session_cookie = session_cookie
        self.base_url = "https://api.graze.social/app/api/v1"
        self.app_base_url = "https://api.graze.social/app"
        self.headers = {
            "Content-Type": "application/json",
            "Origin": "https://www.graze.social",
            "Referer": "https://www.graze.social/",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        }
        
        if session_cookie:
            self.headers["Cookie"] = f"session_cookie={session_cookie}"
    
    def login(self, handle: str, app_password: str) -> Dict[str, Any]:
        """Login to Graze with Bluesky credentials"""
        response = requests.post(
            f"{self.app_base_url}/login",
            json={"username": handle, "password": app_password},
            headers={
                "Content-Type": "application/json",
                "Origin": "https://www.graze.social",
                "Referer": "https://www.graze.social/login",
                "User-Agent": self.headers["User-Agent"]
            }
        )
        
        if response.status_code == 200:
            cookies = response.cookies
            if "session_cookie" in cookies:
                self.session_cookie = cookies["session_cookie"]
                self.headers["Cookie"] = f"session_cookie={self.session_cookie}"
                return {"success": True, "session_cookie": self.session_cookie}
        
        return {"success": False, "error": f"Login failed: {response.status_code}", "response": response.text}
    
    def get_user_id(self) -> str:
        """Extract user ID from session cookie"""
        if not self.session_cookie:
            raise ValueError("No session cookie available")
        
        parts = self.session_cookie.split('.')
        payload = base64.b64decode(parts[0] + '==').decode('utf-8')
        data = json.loads(payload)
        return data["user_id"]
    
    def create_custom_node(self, title: str, description: str, manifest: Dict[str, Any], 
                          color: str = "purple", public: bool = True) -> Dict[str, Any]:
        """Create a new custom node"""
        payload = {
            "name": title,
            "title": title,
            "description": description,
            "public": public,
            "color": color,
            "component": manifest["filter"],
            "component_metadata": {
                "snapshot": {
                    "document": {
                        "store": {},
                        "schema": {"schemaVersion": 2}
                    }
                }
            },
            "version_message": ""
        }
        
        response = requests.post(
            f"{self.base_url}/algorithm-components/components",
            json=payload,
            headers=self.headers
        )
        
        if response.status_code in [200, 201]:
            data = response.json()
            print(f"✓ Created custom node: {title}")
            print(f"  ID: {data.get('id')}")
            print(f"  URL: https://www.graze.social/custom-nodes/{data.get('id')}")
            return data
        else:
            print(f"✗ Failed to create custom node: {response.status_code}")
            print(f"  Response: {response.text}")
            return {"error": response.text, "status_code": response.status_code}
    
    def update_custom_node(self, component_id: str, title: str, description: str, 
                          manifest: Dict[str, Any], color: str = "purple", 
                          public: bool = True) -> Dict[str, Any]:
        """Update an existing custom node"""
        payload = {
            "id": component_id,
            "name": title,
            "title": title,
            "description": description,
            "public": public,
            "color": color,
            "component": manifest["filter"],
            "component_metadata": {
                "snapshot": {
                    "document": {
                        "store": {},
                        "schema": {"schemaVersion": 2}
                    }
                }
            },
            "version_message": ""
        }
        
        response = requests.put(
            f"{self.base_url}/algorithm-components/components/{component_id}",
            json=payload,
            headers=self.headers
        )
        
        if response.status_code == 200:
            data = response.json()
            print(f"✓ Updated custom node: {title}")
            print(f"  URL: https://www.graze.social/custom-nodes/{component_id}")
            return data
        else:
            print(f"✗ Failed to update custom node: {response.status_code}")
            print(f"  Response: {response.text}")
            return {"error": response.text, "status_code": response.status_code}
    
    def get_user_components(self) -> Dict[str, Any]:
        """Get all custom nodes created by the user"""
        user_id = self.get_user_id()
        response = requests.get(
            f"{self.base_url}/algorithm-components/users/{user_id}/components",
            headers=self.headers
        )
        
        if response.status_code == 200:
            return response.json()
        else:
            return {"error": response.text, "status_code": response.status_code}
    
    def push_node_from_file(self, filepath: str, update_id: Optional[str] = None) -> Dict[str, Any]:
        """Push a custom node from a JSON file"""
        with open(filepath, 'r') as f:
            data = json.load(f)
        
        manifest = data.get("manifest", data)
        title = data.get("title", "Untitled Node")
        description = data.get("description", "")
        color = data.get("color", "purple")
        
        if update_id:
            return self.update_custom_node(update_id, title, description, manifest, color)
        else:
            return self.create_custom_node(title, description, manifest, color)


if __name__ == "__main__":
    import sys
    import os
    
    # Example usage
    if len(sys.argv) < 3:
        print("Usage: python graze_client.py <handle> <app_password> [node.json]")
        print("\nExample:")
        print("  python graze_client.py user.bsky.social xxxx-xxxx-xxxx-xxxx my_node.json")
        sys.exit(1)
    
    handle = sys.argv[1]
    app_password = sys.argv[2]
    
    client = GrazeClient()
    
    # Login
    print(f"Logging in as {handle}...")
    result = client.login(handle, app_password)
    
    if not result["success"]:
        print(f"Login failed: {result.get('error')}")
        sys.exit(1)
    
    print("✓ Logged in successfully")
    
    # If node file provided, push it
    if len(sys.argv) > 3:
        node_file = sys.argv[3]
        if os.path.exists(node_file):
            print(f"\nPushing custom node from {node_file}...")
            client.push_node_from_file(node_file)
        else:
            print(f"File not found: {node_file}")
    else:
        print("\nNo node file provided. Use:")
        print(f"  python graze_client.py {handle} <password> node.json")

File: archive/skymap/services/admin/graze-client.js
const axios = require('axios');

class GrazeClient {
  constructor(sessionCookie = null) {
    this.sessionCookie = sessionCookie;
    this.baseURL = 'https://api.graze.social/app/api/v1';
    this.appBaseURL = 'https://api.graze.social/app';
    this.headers = {
      'Content-Type': 'application/json',
      'Origin': 'https://www.graze.social',
      'Referer': 'https://www.graze.social/'
    };
    
    if (sessionCookie) {
      this.headers['Cookie'] = `session_cookie=${sessionCookie}`;
    }
  }

  async login(username, password) {
    try {
      console.log('Making login request to Graze API...');
      
      const response = await axios.post('https://api.graze.social/app/login', {
        username: username,
        password: password
      }, { 
        headers: {
          'Content-Type': 'application/json',
          'Origin': 'https://www.graze.social',
          'Referer': 'https://www.graze.social/login',
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        }
      });
      
      console.log('Login response status:', response.status);
      console.log('Login response data:', response.data);
      
      const cookies = response.headers['set-cookie'];
      if (cookies) {
        console.log('Received cookies:', cookies);
        const sessionCookie = cookies.find(cookie => cookie.startsWith('session_cookie='));
        if (sessionCookie) {
          this.sessionCookie = sessionCookie.split(';')[0].replace('session_cookie=', '');
          this.headers['Cookie'] = `session_cookie=${this.sessionCookie}`;
          return { success: true, sessionCookie: this.sessionCookie };
        }
      }
      
      return { success: false, error: 'No authentication token received' };
    } catch (error) {
      console.error('Login error details:', {
        status: error.response?.status,
        statusText: error.response?.statusText,
        data: error.response?.data,
        message: error.message
      });
      return { success: false, error: JSON.stringify(error.response?.data) || error.message };
    }
  }

  async createCustomNode(title, description, jsonData) {
    try {
      const payload = {
        name: title,
        title: title,
        description: description,
        public: true,
        color: "purple",
        component: jsonData.manifest.filter,
        component_metadata: {
          snapshot: {
            document: {
              store: {},
              schema: {
                schemaVersion: 2
              }
            }
          }
        },
        version_message: ""
      };

      const response = await axios.post(
        `${this.baseURL}/algorithm-components/components`,
        payload,
        { headers: this.headers }
      );

      return response.data;
    } catch (error) {
      console.error('Error creating custom node:', error.response?.data || error.message);
      throw error;
    }
  }

  async updateCustomNode(componentId, title, description, jsonData) {
    try {
      const payload = {
        id: componentId,
        name: title,
        title: title,
        description: description,
        public: true,
        color: "purple",
        component: jsonData.manifest.filter,
        component_metadata: {
          snapshot: {
            document: {
              store: {},
              schema: {
                schemaVersion: 2
              }
            }
          }
        },
        version_message: ""
      };

      const response = await axios.put(
        `${this.baseURL}/algorithm-components/components/${componentId}`,
        payload,
        { headers: this.headers }
      );

      return response.data;
    } catch (error) {
      console.error('Error updating custom node:', error.response?.data || error.message);
      throw error;
    }
  }

  async getUserComponents(userId) {
    try {
      const response = await axios.get(
        `${this.baseURL}/algorithm-components/users/${userId}/components`,
        { headers: this.headers }
      );
      return response.data;
    } catch (error) {
      console.error('Error getting user components:', error.response?.data || error.message);
      throw error;
    }
  }

  async createFeed(displayName, recordName, manifest) {
    try {
      const payload = {
        display_name: displayName,
        record_name: recordName,
        description: "",
        public: false,
        algorithm_type: "hosted"
      };

      const response = await axios.post(
        `${this.appBaseURL}/my_feeds`,
        payload,
        { headers: this.headers }
      );

      return response.data;
    } catch (error) {
      console.error('Error creating feed:', error.response?.data || error.message);
      throw error;
    }
  }

  async getFeed(feedId) {
    try {
      const response = await axios.get(
        `${this.appBaseURL}/publish_algo/${feedId}`,
        { headers: this.headers }
      );
      return response.data;
    } catch (error) {
      console.error('Error getting feed:', error.response?.data || error.message);
      throw error;
    }
  }

  async getFeedContent(feedId, limit = 20, cursor = null) {
    try {
      let url = `${this.appBaseURL}/my_feeds/${feedId}/content?limit=${limit}`;
      if (cursor) {
        url += `&cursor=${encodeURIComponent(cursor)}`;
      }
      
      const response = await axios.get(url, { headers: this.headers });
      return response.data;
    } catch (error) {
      console.error('Error getting feed content:', error.response?.data || error.message);
      throw error;
    }
  }

  async createOrUpdateFeed(displayName, description, stickyType, listUrls, feedId = null, userId = null, publish = false, customManifest = null) {
    try {
      console.log('=== createOrUpdateFeed ===');
      console.log('Display Name:', displayName);
      console.log('Description:', description);
      console.log('Sticky Type:', stickyType);
      console.log('List URLs:', listUrls);
      console.log('Feed ID:', feedId);
      console.log('User ID:', userId);
      console.log('Publish:', publish);
      console.log('Custom Manifest:', !!customManifest);
      
      // Build the manifest
      let manifest;
      if (customManifest) {
        manifest = customManifest;
      } else {
        manifest = {
          filter: {
            or: listUrls.map(url => ({
              list_member: [url, "in"]
            }))
          }
        };
      }

      // Build the algorithm_manifest JSON string
      const algorithmManifest = JSON.stringify(manifest);
      console.log('Algorithm Manifest:', algorithmManifest);

      // Build metadata (empty for now, but matches the structure from your example)
      const metadata = JSON.stringify({
        snapshot: {
          document: {
            store: {},
            schema: { schemaVersion: 2 }
          },
          session: {
            version: 0,
            currentPageId: "page:page",
            exportBackground: true,
            isFocusMode: false,
            isDebugMode: false,
            isToolLocked: false,
            isGridMode: false,
            pageStates: []
          }
        },
        testPosts: []
      });

      // Create FormData
      const FormData = require('form-data');
      const formData = new FormData();
      
      formData.append('display_name', displayName);
      formData.append('description', description || '');
      formData.append('sticky_type', stickyType);
      formData.append('algorithm_manifest', algorithmManifest);
      formData.append('metadata', metadata);
      formData.append('order', stickyType);
      formData.append('active', publish ? 'true' : 'false');
      formData.append('public', publish ? 'true' : 'false');
      if (publish) {
        formData.append('status', 'published');
      }
      
      if (feedId && userId) {
        console.log('Updating existing feed - feedId:', feedId, 'userId:', userId);
        formData.append('user_id', userId.toString());
        formData.append('id', feedId.toString());
      } else if (userId) {
        console.log('Creating new feed with userId:', userId);
        formData.append('user_id', userId.toString());
      } else {
        console.log('Creating new feed without userId');
      }

      const headers = {
        ...this.headers,
        ...formData.getHeaders()
      };
      delete headers['Content-Type']; // Let form-data set the correct boundary

      const endpoint = feedId 
        ? `${this.appBaseURL}/edit_algo_image`
        : `${this.appBaseURL}/add_algo_image`;

      console.log('Endpoint:', endpoint);
      console.log('Headers:', JSON.stringify(headers, null, 2));

      const response = await axios.post(endpoint, formData, { headers });
      
      console.log('Response status:', response.status);
      console.log('Response data:', JSON.stringify(response.data, null, 2));

      return response.data;
    } catch (error) {
      console.error('Error in createOrUpdateFeed:', error.message);
      if (error.response) {
        console.error('Response status:', error.response.status);
        console.error('Response data:', JSON.stringify(error.response.data, null, 2));
      }
      throw error;
    }
  }

  async getUserId() {
    try {
      // Extract user_id from the session cookie
      // Cookie format: session_cookie=eyJzZXNzaW9uX3Rva2VuIjoiLi4uIiwidXNlcl9pZCI6NTk1M30...
      const cookieValue = this.sessionCookie;
      if (!cookieValue) {
        throw new Error('No session cookie');
      }
      
      // Decode the base64 JWT payload (middle part)
      const parts = cookieValue.split('.');
      if (parts.length < 2) {
        throw new Error('Invalid session cookie format');
      }
      
      const payload = Buffer.from(parts[0], 'base64').toString('utf8');
      const data = JSON.parse(payload);
      
      console.log('Decoded session data:', data);
      
      if (data.user_id) {
        return data.user_id;
      }
      
      throw new Error('user_id not found in session');
    } catch (error) {
      console.error('Error getting user ID:', error.message);
      throw error;
    }
  }

  async publishFeed(feedId) {
    try {
      console.log('Publishing feed via GET:', feedId);
      
      const response = await axios.get(
        `${this.appBaseURL}/publish_algo/${feedId}`,
        { headers: this.headers }
      );
      
      console.log('Publish response:', response.data);
      return response.data;
    } catch (error) {
      console.error('Error publishing feed:', error.response?.data || error.message);
      throw error;
    }
  }
}

module.exports = GrazeClient;